### PR TITLE
[SCEV] Use context sensitive reasoning in howFarToZero

### PR DESCRIPTION
**Pull request by preames**
This change builds on 0a357ad which supported non-constant strides in howFarToZero, but used only context insensitive reasoning.

This change does two things:
1) Directly use context sensitive queries to prove facts established
   before the loop.  Note that we technically only need facts known
   at the latch, but using facts known on entry is a conservative
   approximation which will cover most everything.
2) For the non-zero check, we can usually prove non-zero from the
   finite assumption implied by mustprogress.  This eliminates the
   need to do the context sensitive query in the common case.

### CODE REVIEW

**Comment by preames** [comment]
Looking at this code, I think we can further improve the reasoning here in a couple of ways.  Let me sketch out a few ideas here, so reviewers can see where this is going.

* I think we can avoid the need to statically know the sign bit.  I think I have a SCEV expression which represents the select required for the Distance - though this is pending verification.
* I think we're being too conservative about applying no-self-wrap.  (And by "we", I mean "I" as I wrote that code...)  If we have any (not just sole) latch dominating exit which is controlled by an NW addrec, then we'd branch-on-poison on the iteration which violates the NW.  As such, we can assume that that some other exit is taken first.  We can consider the divide (ignoring wrap) as a bound for this exit by simply adding one to the result.
* Looking at other exit computation, it looks like I got the nw property wrong (too conservative) in a few cases, in slightly different ways.  It looks like maybe I confused myself between the actual distance bound, and the repeats the same value variant - which aren't quite the same.  

I'm probably going to post a patch for the first of those.  Not sure I'm going to continue through with the later two, as they're a bit more involved.

**Review by nikic — COMMENTED** [review]
This one does have some overhead: https://llvm-compile-time-tracker.com/compare.php?from=21711f89b9d85028160611f725bd33d7832d1d46&to=d89246df52708a8fd1fb126526f104ac89240fef&stat=instructions:u

It would be better to use applyLoopGuards, which should be cheaper and also provide better results for the queries you are performing here.

**Comment by preames** [comment]
> It would be better to use applyLoopGuards, which should be cheaper and also provide better results for the queries you are performing here.

How would loop guards help prove the

**Comment by preames** [comment]
@nikic ping

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 1]
```diff
@@ -10485,16 +10485,19 @@ ScalarEvolution::ExitLimit ScalarEvolution::howFarToZero(const SCEV *V,
   const SCEV *Step = getSCEVAtScope(AddRec->getOperand(1), L->getParentLoop());
   const SCEVConstant *StepC = dyn_cast<SCEVConstant>(Step);
 
-  if (!isLoopInvariant(Step, L) || !isKnownNonZero(Step))
+  if (!isLoopInvariant(Step, L))
     return getCouldNotCompute();
 
+  // Specialize step for this loop so we get context sensative facts below.
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:10491]:
```suggestion
  // Specialize step for this loop so we get context sensitive facts below.
```

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 2]
```diff
@@ -10533,6 +10536,13 @@ ScalarEvolution::ExitLimit ScalarEvolution::howFarToZero(const SCEV *V,
   // will have undefined behavior due to wrapping.
   if (ControlsOnlyExit && AddRec->hasNoSelfWrap() &&
       loopHasNoAbnormalExits(AddRec->getLoop())) {
+
+    // If the stride is zero, the loop must be infinite.  Most loops are
+    // finite by assumption, in which case the step being zero implies UB
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:10541]: "Most loops are finite by assumption" is a C++-ism...

**Review by nikic — APPROVED** [review]
Unfortunately the overhead stays about the same: http://llvm-compile-time-tracker.com/compare.php?from=21711f89b9d85028160611f725bd33d7832d1d46&to=234423fe3d82181c2cc1a7a614f60b46dea70297&stat=instructions%3Au

But it's small enough that I'm okay with landing this if it's useful.

**Pull request merged**